mender-update/daemon/context.cpp

```diff
@@ -142,6 +146,299 @@ Context::Context(main_context::MenderContext &mender_context, events::EventLoop
+
+	if (state_data.update_info.state_data_store_count++ >= kMaxStateDataStoreCount) {
+		return main_context::MakeError(
+			main_context::StateDataStoreCountExceededError,
+			"State looping detected, breaking out of loop");
+	}
+
+	string content = GenerateStateDataJson(state_data);
+
+	// TODO: Handle commit stage here.
+	string store_key;
+	if (state_data.update_info.has_db_schema_update) {
+		store_key = mender_context.state_data_key_uncommitted;
+	} else {
+		store_key = mender_context.state_data_key;
+	}
+
+	auto err = txn.Write(store_key, vector<uint8_t>(content.begin(), content.end()));
+	if (err != error::NoError) {
+		return err.WithContext("Could not write state data");
+	}
+
+	return error::NoError;
+}
+
+error::Error Context::SaveDeploymentStateData(StateData &state_data) {
+	auto &db = mender_context.GetMenderStoreDB();
+	return db.WriteTransaction([this, &state_data](kv_db::Transaction &txn) {
+		return SaveDeploymentStateData(txn, state_data);
+	});
+}
+
+static error::Error UnmarshalJsonStateData(const json::Json &json, StateData &state_data) {
+#define SetOrReturnIfError(dst, expr) \
+	if (!expr) {                      \
+		return expr.error();          \
+	}                                 \
+	dst = expr.value()
+
+#define EmptyOrSetOrReturnIfError(dst, expr)                                   \
+	if (!expr) {                                                               \
+		if (expr.error().code == kv_db::MakeError(kv_db::KeyError, "").code) { \
+			dst.clear();                                                       \
+		} else {                                                               \
+			return expr.error();                                               \
+		}                                                                      \
+	} else {                                                                   \
+		dst = expr.value();                                                    \
+	}
+
+	auto get_vector = [](const json::Json &json) -> expected::ExpectedStringVector {
+		auto exp_size = json.GetArraySize();
+		if (!exp_size) {
+			return expected::unexpected(exp_size.error());
+		}
+		auto size = exp_size.value();
+
+		vector<string> ret;
+		for (size_t index = 0; index < size; index++) {
+			auto exp_string = json[index].and_then(json::ToString);
+			if (!exp_string) {
+				return expected::unexpected(exp_string.error());
+			}
+			ret.push_back(exp_string.value());
+		}
+
+		return ret;
+	};
+
+	auto get_map = [](const json::Json &json) -> json::ExpectedKeyValueMap {
+		auto exp_children = json.GetChildren();
+		if (!exp_children) {
+			return expected::unexpected(exp_children.error());
+		}
+		const auto &children = exp_children.value();
+
+		unordered_map<string, string> ret;
+		for (auto &elem : children) {
+			auto exp_string = elem.second.GetString();
+			if (!exp_string) {
+				return expected::unexpected(exp_string.error());
+			}
+			ret[elem.first] = exp_string.value();
+		}
+
+		return ret;
+	};
```

Comment:
`json::ToKeyValuesMap()`

Reply:
How on earth did I miss that?? 😮 Especially since I added other functions to help me!